ReactiveCocoa/Swift/Signal.swift

```diff
@@ -243,6 +243,324 @@ extension SignalType {
 	}
 }
 
+/// Describes how multiple producers should be joined together.
+public enum FlattenStrategy: Equatable {
+	/// The producers should be merged, so that any value received on any of the
+	/// input producers will be forwarded immediately to the output producer.
+	///
+	/// The resulting producer will complete only when all inputs have completed.
+	case Merge
+
+	/// The producers should be concatenated, so that their values are sent in the
+	/// order of the producers themselves.
+	///
+	/// The resulting producer will complete only when all inputs have completed.
+	case Concat
+
+	/// Only the events from the latest input producer should be considered for
+	/// the output. Any producers received before that point will be disposed of.
+	///
+	/// The resulting producer will complete only when the producer-of-producers and
+	/// the latest producer has completed.
+	case Latest
+}
+
+extension Signal where T: SignalProducerType, E == T.E {
+	/// Flattens the inner producers sent upon `signal` (into a single signal of
+	/// values), according to the semantics of the given strategy.
+	///
+	/// If `signal` or an active inner producer emits an error, the returned
+	/// signal will forward that error immediately.
+	///
+	/// `Interrupted` events on inner producers will be treated like `Completed`
+	/// events on inner producers.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	public func flatten(strategy: FlattenStrategy) -> Signal<T.T, E> {
+		switch strategy {
+		case .Merge:
+			return signal.merge()
+
+		case .Concat:
+			return signal.concat()
+
+		case .Latest:
+			return signal.switchToLatest()
+		}
+	}
+}
+
+extension Signal where T: SignalProducerType, E == T.E {
+	/// Returns a signal which sends all the values from producer signal emitted from
+	/// `signal`, waiting until each inner producer completes before beginning to
+	/// send the values from the next inner producer.
+	///
+	/// If any of the inner producers emit an error, the returned signal will emit
+	/// that error.
+	///
+	/// The returned signal completes only when `signal` and all producers
+	/// emitted from `signal` complete.
+	@warn_unused_result(message="Did you forget to call `observe` on the signal?")
+	private func concat() -> Signal<T.T, E> {
+		return Signal<T.T, E> { [signal = self.signal] observer in
+			let disposable = CompositeDisposable()
+			let state = ConcatState(observer: observer, disposable: disposable)
+
+			disposable += signal.observe { event in
+				switch event {
+				case .Next(let value):
+					state.enqueueSignalProducer(value.producer)
+					break
+
+				case .Error(let error):
+					sendError(observer, error)
+
+				case .Completed:
+					// Add one last producer to the queue, whose sole job is to
+					// "turn out the lights" by completing `observer`.
+					let completion: SignalProducer<T.T, E> = SignalProducer<T.T, E>
```

Comment:
Actually, can you get rid of the local and pass this directly to `enqueueSignalProducer`?